### ISSUE TITLE
"Invalid X errors" should indicate what the valid format is

### ISSUE THREAD

**Issue by MariaSolOs** [nushell/nushell#8927]
### Related problem

Recently I was trying to implement a custom command and wondered what the right syntax for a multi-word parameter was. Since I was following the [documentation about custom commands](https://www.nushell.sh/book/custom_commands.html#command-names), I guessed that I should use a dash since "it's common practice in Nushell to separate the words of the command with - for better readability". So I tried the following:
```nu
export def foo [bar-bar: string] {
    print $bar-bar
}
```

but this resulted in:
```
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[test.nu:1:1]
 1 │ export def foo [bar-bar: string] {
   ·                 ────┬───
   ·                     ╰── expected valid variable name for this parameter
 2 │     print $bar-bar
   ╰────
```

I was then stuck for a bit, because I couldn't find anything in the docs about what a valid variable name was. I figured it out by matching the parse error to the [Nushell source code](https://github.com/nushell/nushell/blob/91c01bf6b3bc58fc1510ba0caa3da414b76fac36/crates/nu-parser/src/parse_patterns.rs#L77), but that's not a great user experience :(

### Describe the solution you'd like

The diagnostic should include further information about what a valid variable name is. This potentially applies to other "invalid X" errors (I haven't checked if other diagnostics have the same problem).

**Comment by amtoine** [comment]
> ```nu
> export def foo [$bar-bar: string] {
>     print $bar-bar
> }
> ```
>
yeah it should be something like
```nushell
export def foo [bar_bar: string] {
    print $bar_bar
}
```

> The diagnostic should include further information about what a valid variable name is. This potentially applies to other "invalid X" errors (I haven't checked if other diagnostics have the same problem).
>
better error sounds sensible to me :yum:
at least a quick fix and a link to the book for further information :thumbsup:

**Comment by MariaSolOs** [comment]
@amtoine agreed, although I think that the book doesn't have any information about valid variable names (?). All I could find was:
![image](https://user-images.githubusercontent.com/62502207/233221429-8230486b-57da-493f-b297-8629eb1a3ca1.png)

which is clearly not applicable here.

Also I realized that I had a typo in the original description. The parameter wasn't supposed to have a `$` (so only the dashes were supposed to be invalid).

**Comment by amtoine** [comment]
> @amtoine agreed, although I think that the book doesn't have any information about valid variable names (?).
> 
i'm not sure, it could be missing :thinking: 
did you search for "variable" or "variable name" or "naming convention"?

> Also I realized that I had a typo in the original description. The parameter wasn't supposed to have a `$` (so only the dashes were supposed to be invalid).
>
funny enough
```
def foo [$x: string] { print $x }
```
appears to work :scream: :wink:

**Comment by amtoine** [comment]
@MariaSolOs 
i completely agree with you, it's quite important :thinking: 

do you have time and energy for this? :yum:

**Comment by MariaSolOs** [comment]
> @MariaSolOs i completely agree with you, it's quite important 🤔
> 
> do you have time and energy for this? 😋

Of course!

**Comment by amtoine** [comment]
well then, if you want a default reviewer, you know where to ping... :wink:

**Comment by 1Kinoti** [comment]
@MariaSolOs here's the [updated](https://www.nushell.sh/book/variables_and_subexpressions.html#variable-names) docs on variable names

**Comment by MariaSolOs** [comment]
@1Kinoti Thanks a lot! I took a look at them and I left some suggestions in https://github.com/nushell/nushell.github.io/pull/881 😺
